### PR TITLE
Fixes for Issues #162 and #164; Add more metrics for PageDeviceCache and PageCacheSlot::Pool

### PR DESCRIPTION
**Pull request by vidyasilai**
- Fixes for the bugs mentioned in [issue 162](https://github.com/mathworks/llfs/issues/162) and [issue 164](https://github.com/mathworks/llfs/issues/164).
- Incorporated the metrics defined in page_cache_slot_pool.hpp into PageCacheDevice and PageCacheSlot::Pool classes.

### CODE REVIEW

**Review by tonyastolfi — COMMENTED** [review]
Looks great overall!  I think there is an unhandled case in `evict_if_key_equals`, and I left a few other minor comments.  Also, please consider adding a multi-threaded stress test that calls `acquire_pin`/`release_pin` concurrent to `evict_if_key_equals` (shouldn't be too hard with a focused test to trigger the failed check for #164).

File: src/llfs/page_cache_slot_pool.hpp [hunk 1]
```diff
@@ -93,7 +93,7 @@ class PageCacheSlot::Pool : public boost::intrusive_ref_counter<Pool>
 
   /** \brief Returns the metrics for this pool.
    */
-  const Metrics& metrics() const
+  Metrics& metrics()
```
[the remark this answers is not in the document]
Comment (vidyasilai) [reply src/llfs/page_cache_slot_pool.hpp:103]: Added back, and done the same for the comment below

File: src/llfs/page_cache_slot.cpp [hunk 2]
```diff
@@ -258,6 +272,7 @@ bool PageCacheSlot::evict_if_key_equals(PageId key) noexcept
 
     if (this->state_.compare_exchange_weak(observed_state, target_state)) {
       BATT_CHECK(!Self::is_valid());
+      this->remove_ref();
```
[the remark this answers is not in the document]
Comment (vidyasilai) [reply src/llfs/page_cache_slot.cpp:279]: Added comment

File: src/llfs/page_cache_slot.cpp [hunk 3]
```diff
@@ -235,6 +244,11 @@ bool PageCacheSlot::evict_if_key_equals(PageId key) noexcept
...
   auto observed_state = old_state + kPinCountDelta;
 
+  const bool newly_pinned = !this->is_pinned(old_state);
+  if (newly_pinned) {
+    this->add_ref();
+  }
+
   BATT_CHECK_EQ(observed_state & Self::kOverflowMask, 0);
 
   // Use a CAS loop here to guarantee an atomic transition from Valid + Filled (unpinned) state to
```
[the remark this answers is not in the document]
Comment (vidyasilai) [reply src/llfs/page_cache_slot.cpp:254]: Changed the `fetch_sub` to a call to `release_pin`

**Review by tonyastolfi — APPROVED** [review]
LGTM!

**Pull request merged**